dash-spv-apple-bindings/src/coinjoin.rs

```diff
@@ -38,44 +34,65 @@ pub unsafe extern "C" fn register_coinjoin(
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn register_wallet_ex(
+pub unsafe extern "C" fn register_client_session(
+    coinjoin: *mut CoinJoin,
     options_ptr: *mut CoinJoinClientOptions,
     get_wallet_transaction: GetWalletTransaction,
     destroy_wallet_transaction: DestroyWalletTransaction,
     is_mine: IsMineInput,
     has_collateral_inputs: HasCollateralInputs,
     selected_coins: SelectCoinsGroupedByAddresses,
     destroy_selected_coins: DestroySelectedCoins,
+    sign_transaction: SignTransaction,
     context: *const c_void
-) -> *mut WalletEx {
-    let options: CoinJoinClientOptions = std::ptr::read(options_ptr);
-    let wallet_ex = WalletEx::new(
-        context, 
-        options, 
+) -> *mut CoinJoinClientSession {
+    let session = CoinJoinClientSession::new(
+        std::ptr::read(coinjoin),
+        std::ptr::read(options_ptr), 
+        sign_transaction,
         get_wallet_transaction, 
         destroy_wallet_transaction, 
         is_mine, 
         has_collateral_inputs, 
         selected_coins, 
-        destroy_selected_coins
+        destroy_selected_coins,
+        context
     );
-    println!("[RUST] register_wallet_ex: {:?}", wallet_ex);
-    boxed(wallet_ex)
+    println!("[RUST] register_session");
+    boxed(session)
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn unregister_coinjoin(coinjoin: *mut CoinJoin) {
-    println!("[RUST] 💀 unregister_coinjoin: {:?}", coinjoin);
-    let unboxed = unbox_any(coinjoin);
+pub unsafe extern "C" fn call_session(
+    session: *mut CoinJoinClientSession
+) -> bool {
+    println!("[RUST] call session");
+    (*session).run_session();
+    return true;
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn call_coinjoin(
-    coinjoin: *mut CoinJoin
-) -> bool {
-    println!("[RUST] call coinjoin");
-    // return (*coinjoin).trigger_session();
-    return true;
+pub unsafe extern "C" fn on_transaction_signed_for_session(
+    tx: *mut types::Transaction, 
+    session: *mut CoinJoinClientSession,
+    destroy_wallet_transaction: DestroyWalletTransaction
+) {
+    (*session).on_transaction_signed((*tx).decode());
+    (destroy_wallet_transaction)(tx);
+}
```

Comment:
The thing I got stuck on is how to sign a transaction given that in DashSync, the `sign` method returns the result in a callback.
I tried to pass around a pointer for the `TransactionBuilder` so that it's called here, but it proved challenging due to the Rust memory management. In the end, I ended up saving the current `TransactionBuilder` as one of the Sessions fields. The session itself is kept alive by DashSync.

Perhaps the easy way would be to have a `sign` method without a callback (since we can now sign transactions without showing a prompt to the user anyway).

Otherwise, maybe there is some simple way to handle a callback in Rust that I'm missing.

Reply:
Which method "sign" are you referring to?

Reply:
`signTransaction` in DashSync: https://github.com/dashpay/dashsync-iOS/blob/master/DashSync/shared/Models/Wallet/DSAccount.h#L182

Reply:
“Pre-ferment” way of doing that is 

Rust:

- Make a struct with callbacks like [this](https://github.com/dashpay/dash-shared-core/blob/8a8c03caad1a39965cb4f426c2796a72e241992d/dash-spv-masternode-processor/src/bindings/common.rs#L53) (in ferment there will be likely one-callback-per-struct)
- Make bindings for alloc/dealloc where you deal with opaque pointer for this struct like [this](https://github.com/dashpay/dash-shared-core/blob/8a8c03caad1a39965cb4f426c2796a72e241992d/dash-spv-masternode-processor/src/bindings/common.rs#L53C1-L54C1) and [that](https://github.com/dashpay/dash-shared-core/blob/8a8c03caad1a39965cb4f426c2796a72e241992d/dash-spv-masternode-processor/src/bindings/common.rs#L92)
- Whenever you use a specific binding like [this](https://github.com/dashpay/dash-shared-core/blob/8a8c03caad1a39965cb4f426c2796a72e241992d/dash-spv-masternode-processor/src/bindings/masternode.rs#L17) you’re passing this opaque context

ObjC

- [Brother-in-law for MasternodeProcessor](https://github.com/dashpay/dashsync-iOS/blob/080be301d48a64e29df93855fe9974eb616ab8f5/DashSync/shared/Models/Managers/Chain%20Managers/DSMasternodeManager.m#L70C17-L70C36) [alloc](https://github.com/dashpay/dashsync-iOS/blob/080be301d48a64e29df93855fe9974eb616ab8f5/DashSync/shared/Models/Managers/Chain%20Managers/DSMasternodeManager.m#L95C1-L97C1) and [dealloc](https://github.com/dashpay/dashsync-iOS/blob/080be301d48a64e29df93855fe9974eb616ab8f5/DashSync/shared/Models/Managers/Chain%20Managers/DSMasternodeManager.m#L72) this pointer
- [Binding](https://github.com/dashpay/dashsync-iOS/blob/080be301d48a64e29df93855fe9974eb616ab8f5/DashSync/shared/Models/Managers/Chain%20Managers/DSMasternodeManager%2BMndiff.m#L247) to pass the opaque pointer
- Callbacks maintenance like [this](https://github.com/dashpay/dashsync-iOS/blob/080be301d48a64e29df93855fe9974eb616ab8f5/DashSync/shared/Models/Managers/Chain%20Managers/DSMasternodeManager%2BMndiff.m#L32)

Strategically, It’s better to have a single opaque context for entire rust sdk lib. 

For now I think it’s feasible to have multiple before we come up with a more solid rust spv sdk with chain support

Reply:
I'll take a look at how to handle signing when it comes up again